internal/serialization/portable_serializer_test.go

```diff
@@ -528,5 +529,128 @@ func TestPortableSerializer_NestedPortableVersion(t *testing.T) {
 	if !reflect.DeepEqual(deserializedParent, p) {
 		t.Error("nested portable version is wrong")
 	}
+}
+
+const (
+	factoryID1 = 1
+	factoryID2 = 2
+)
+
+type MyPortable1 struct {
+	stringField string
+}
+
+func (MyPortable1) FactoryID() (factoryID int32) {
+	return factoryID1
+}
+
+func (MyPortable1) ClassID() (classID int32) {
+	return 1
+}
+
+func (p MyPortable1) WritePortable(writer serialization.PortableWriter) {
+	writer.WriteString("stringField", p.stringField)
+}
+
+func (p *MyPortable1) ReadPortable(reader serialization.PortableReader) {
+	p.stringField = reader.ReadString("stringField")
+}
+
+type MyPortable2 struct {
+	intField int32
+}
+
+func (MyPortable2) FactoryID() (factoryID int32) {
+	return factoryID2
+}
+
+func (MyPortable2) ClassID() (classID int32) {
+	return 1
+}
+
+func (p MyPortable2) WritePortable(writer serialization.PortableWriter) {
+	writer.WriteInt32("intField", p.intField)
+}
+
+func (p *MyPortable2) ReadPortable(reader serialization.PortableReader) {
+	p.intField = reader.ReadInt32("intField")
+}
+
+type MyPortableFactory1 struct {
+}
+
+func (MyPortableFactory1) Create(classID int32) (instance serialization.Portable) {
+	if classID == 1 {
+		return &MyPortable1{}
+	}
+	return
+}
+
+func (MyPortableFactory1) FactoryID() int32 {
+	return factoryID1
+}
+
+type MyPortableFactory2 struct {
+}
+
+func (MyPortableFactory2) Create(classID int32) (instance serialization.Portable) {
+	if classID == 1 {
+		return &MyPortable2{}
+	}
+	return
+}
+
+func (*MyPortableFactory2) FactoryID() int32 {
+	return factoryID2
+}
+
+func TestClassesWithSameClassIdInDifferentFactories(t *testing.T) {
+	it.MarkFlaky(t, "does not pass")
+	config := &serialization.Config{}
+	config.PortableVersion = 3
```

Comment:
I can't see the corresponding line for this in the Java client test.

Reply:
Default portable version is zero for non-versioned classes. We can override that version. `NewService` use different  portable version (default: 0) in that case although I specify portable as 3. That can create a problem. You are right.
Java has both:
- `public ClassDefinitionBuilder(int factoryId, int classId)` -> Default is being used.
- `public ClassDefinitionBuilder(int factoryId, int classId, int version)` -> given portable version should match with the portable version in serialization service.
**But we have only a `NewClassDefinition` which does take version as well like in the second one in java. That's why I used it. We can add non-versioned one as well in public api.**

**Another suggestion to prevent miss usage:**
We can ensure whether given version through class definition in `SetClassDefinitions` is equivalent with the portable version in the serialization service. It can prevent user to give non-matched portable version 🤔
Like this: [cd443da](https://github.com/hazelcast/hazelcast-go-client/pull/758/commits/cd443da556767f116d96dd7a518151744bad043e).

Reply:
I don't want to add a new public API for portable stuff, unless we must. Could you add some documentation to `NewClassDefinition` which mentions, in order to use the default version it should be called with `version == 0` ?

Reply:
Added, [5dca40b](https://github.com/hazelcast/hazelcast-go-client/pull/758/commits/5dca40bd3b2c69bfe4310bb3c293c71d358b2116)